open_spiel/games/2048.cc

```diff
@@ -0,0 +1,382 @@
+  int count = AvailableCellCount();
+  if (count == 0) {
+    return {{kNoCellAvailableAction, 1.0}};    
+  }
+  ActionsAndProbs action_and_probs;
+  action_and_probs.reserve(count * 2);
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      if (BoardAt(r, c).value == 0) {
+        // 2 appearing randomly on the board should be 9 times as likely as a 4
+        action_and_probs.emplace_back(ChanceActionToSpielAction(
+            ChanceAction(r, c, false)), .9 / count);
+        action_and_probs.emplace_back(ChanceActionToSpielAction(
+            ChanceAction(r, c, true)), .1 / count);
+      }      
+    }
+  }  
+  return action_and_probs;
+}
+
+std::vector<Action> TwoZeroFourEightState::LegalActions() const {
+  if (IsTerminal()) {
+    return {};
+  }
+  if (IsChanceNode()) {
+    return LegalChanceOutcomes();
+  }
+  return kPlayerActions();
+}
+
+bool TwoZeroFourEightState::InBounds(int row, int column) const {
+  return (row >= 0 && row < kDefaultRows && column >= 0
+      && column < kDefaultColumns);
+}
+
+std::string TwoZeroFourEightState::ToString() const {  
+  std::string str;
+  for (int r = 0; r < kDefaultRows; ++r) {
+    for (int c = 0; c < kDefaultColumns; ++c) {
+      std::string tile = std::to_string(BoardAt(r, c).value);
+      absl::StrAppend(&str, std::string(5 - tile.length(), ' '));
+      absl::StrAppend(&str, tile);
+    }
+    absl::StrAppend(&str, "\n");
+  }
+  return str;  
+}
+
+bool TwoZeroFourEightState::IsTerminal() const {
+  return Reached2048() 
+      || (AvailableCellCount() == 0 && !TileMatchesAvailable());
+}
+
+bool TwoZeroFourEightState::Reached2048() const {
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      if (BoardAt(r, c).value == 2048) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+std::vector<double> TwoZeroFourEightState::Rewards() const {
+  return {action_score_};
+}
+
+std::vector<double> TwoZeroFourEightState::Returns() const {
+  return {total_score_};
+}
+
+std::string TwoZeroFourEightState::InformationStateString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return HistoryString();
+}
+
+std::string TwoZeroFourEightState::ObservationString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return ToString();
+}
+
+void TwoZeroFourEightState::ObservationTensor(Player player,
+                                      absl::Span<float> values) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  TensorView<2> view(values, {kDefaultRows, kDefaultColumns}, true);
+  for (int row = 0; row < kDefaultRows; row++) {
+    for (int column = 0; column < kDefaultColumns; column++) {
+      view[{row, column}] = BoardAt(row, column).value;
+    }
+  }
+}
+
+void TwoZeroFourEightState::UndoAction(Player player, Action action) {  
+  history_.pop_back();
```

Comment:
I was thinking of not implementing UndoAction if it's not necessary for all games in OpenSpiel. I can remove this function if that's the case.

Otherwise I think I'll just have to keep a copy of the previous state as reconstructing the state from the move played and other associated info can get quite complex